core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -322,21 +379,51 @@ private Object maybeUpdate(String hostname, CacheEntry entry) {
         }
 
         boolean refreshable() {
-            return address != null;
+            if (address == null) {
+                return false;
+            }
+
+            if (autoRefreshTimeoutFunction == DEFAULT_AUTO_REFRESH_TIMEOUT_FUNCTION) {
+                return true;
+            }
+
+            try {
+                final long timeoutMillis = autoRefreshTimeoutFunction.applyAsLong(hostname);
+                if (timeoutMillis == Long.MAX_VALUE) {
+                    return true;
+                }
+                if (timeoutMillis <= 0) {
+                    return false;
+                }
+
+                final long elapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - creationTimeNanos);
```

Comment:
Shouldn't we rename `creationTimeNanos` and update it whenever the cache hits?

Reply:
As jrhee17 said, `creationTimeNanos` is not updated. A new refreshed `CacheEntry` inherits the original `creationTimeNanos` to calculate the refresh timeout.
https://github.com/line/armeria/blob/6aa51c2f6cb81b36f3be79985b4fa065d1808789/core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java#L338
The name is changed to `originalCreattionTimeNanos` for clarity.

Reply:
> I actually understood this to be an equivalent of maxAge of CacheEntry

If it's the `maxAge`, we don't have to calculate every time when it's removed but we can calculate the deadline one time in the constructor and check if it's passed or not in `refreshable()`.

Reply:
Techinically, `originalCreationTimeNanos` is different from `maxAge`.
`maxAge` is close to the timeout of a `CacheEntry`. But `originalCreationTimeNanos` is when the `CacheEntry` is created which is used to compute the remaining time to predicate the `CacheEntry` is expired.